scripts/changelog-check

```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env ruby
+require 'open3'
+require 'optparse'
+
+# [skip_changelog]
+
+CHANGELOG_REGEX = %r{^\+- (?<category>[\w ]+{2,}): (?<change>[\w\s]+{2,})}
+# A valid diff line is in the form of:
+# +- CATEGORY: CHANGE DESCRIPTION
+def is_valid_changelog_diff_addition?(line)
+  matches = CHANGELOG_REGEX.match?(line)
+  !matches.nil?
```

Comment:
no need for the intermediate, right?
```suggestion
  CHANGELOG_REGEX.match?(line)
```

or maybe add a `!!` if we want the boolean instead of nil-ish

```suggestion
  !!CHANGELOG_REGEX.match?(line)
```

Reply:
The intermediate part isn't used yet, but I'd like to keep it because the match capture groups contain the category and change description. We aren't using them yet, but it could be useful if we'd like to do validation on those in the future.

Reply:
Actually, there may be some validation we want included in the changelog description to start with like ending with PR number. What do you think?

Reply:
Yeah so like in this example:
```
- Add CI check to include changelog message in change requests (#5836)
```
is missing a category like `Source Code: ` or something right?

If it helps us make a better error message, I'm all for keeping it

Reply:
Yeah, exactly. I'll add the validation to include `(#5836)`, and keep the category validation just for existence for now. If we want to have a set of categories we want to limit ourselves to at some point, that's an option.

Reply:
I wish there was a way to auto-add the PR number!

Reply:
Yeah, I don't like that it's a chicken/egg thing, it'd be nice to have a good solution for that.

Reply:
If the idea is that we could automate compiling the CHANGELOG notes from the git/GitHub history, could we invert it so that the pull request number is added in that notes compilation step? That might also open up the possibility of having multiple pull requests correspond to the same change item.

Reply:
My aim was to be able to copy/paste the markdown from the CHANGELOG into the release notes without much/any tooling in between. I don't like the potential friction where a pull request could be opened and be completely valid with the exception of not being able to know the PR number ahead of time for the changelog and failing because of that.

I'm not sure, I lean towards just not requiring the PR number and having the release manager do it manually, though I guess it would then also require an update to the changelog. A bit torn here.

The multiple PRs support required a small regex change in da7969c.

Reply:
Oh, I missed that the pull request author would be adding to `CHANGELOG.md` directly. I think the way I assumed it would work is that a CHANGELOG note would be included in either the commit message or the pull request description, and then we could generate a CHANGELOG from that (maybe added to `CHANGELOG.md` in the stage branch). I'm not sure which is better, but I do still have some concern about merge conflicts, and then yes, it would be quite a burden to have a guaranteed build failure for a new pull request until the number is added.

Reply:
Hm I like the commit message approach, it would make the PR process easier (since you wouldn't have to amend) but it would require separate tooling to combine the PRs in a release into release notes

At one point in time we had some git hooks installed via overcommit that would warn if commit messages didn't have `**Why**` so I think it's possible to make PRs fail in a similar way (#4123) (I don't think git hooks are the right way but it's an idea)

Reply:
Yeah, without a good option for `CHANGELOG.md`, I'm leaning towards experimenting with something else. Git commit messages would be good, I would like to use [trailers](https://git-scm.com/docs/git-interpret-trailers) as GitLab describes [here](https://docs.gitlab.com/ee/development/changelog.html), but that only works if it's the absolute end of the commit message, which probably isn't feasible due to our squashing of PRs.

Reply:
Oh I just figured we'd come up with a magic format like this:

```
RELEASE_NOTES: Some Category: Pull Requests now require release notes
```

So as long as one of the commit messages contains it, it would be fine (I typically clean my commits up before squashing but 🤷 ), and then we rely on GH putting the PR number in the first line, so it would be like:

```
My Pull Request Title (#123)

Updates the codebase to do some thing

RELEASE_NOTES: Some Category: Pull Requests now require
release notes and line breaks are okay because we're smart now

* Second commit message from squashing

* Third commit message
```

Reply:
Yeah, that makes sense and I'll give that a shot. Trailers are very much in the nice-to-have category, not strictly required.